### PR TITLE
Fix callbacks not being fired on macOS ActionSheetIOS

### LINKED ISSUE #1680: Refactor RCTActionSheetManager macOS implementation to not rely on removed `_callbacks`
React Native Core no longer uses it for iOS as of https://github.com/facebook/react-native/pull/33097 but the macOS implementation still has one call to it. Can we refactor and remove it to remove a diff?

### PR DESCRIPTION
**Pull request by Saadnajmi**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary
Resolves #1680 
On macOS, we implement ActionSheetIOS to be an NSMenu (not confusing at all). During the 0.71 merge, `RCTActionSheetManager.mm` was refactored a bit, and the macOS implementation wasn't properly updated. Thus, callbacks wouldn't get fired when you clicked on menu items.

While the fix is simple (initialize `_callbacks` on init), I decided to go ahead and refactor the diffs so that the iOS and macOS blocks are much closer to each other and share more code. I also added a few comments explaining some macOS differences we needed. This should help ensure that future merges are less likely to break this module.

Things to note:
- This PR also fixes the ability to disable menu items, which previously didn't work. Yay code sharing :D.
- NSMenu doesn't have a "cancel" menu item like iOS Action Sheet buttons do, you just click away to close the menu. So, we just don't add a menu item for any button marked as "cancel"
- NSMenuItems don't have a "destructive" style like iOS Action Sheet buttons do. So we don't do any special styling for them.
- I didn't refactor any of the Share dialog bit because there isn't a bug there, and I decided to leave as is.

## Changeling

[MACOS] [FIXED] - Fix callbacks not being fired on macOS ActionSheetIOS

## Test Plan

See video. Before, clicking any menu item would result in a warning that no callback is registered.

[Screen Recording 2023-05-25 at 12.34.53 PM](https://github.com/microsoft/react-native-macos/assets/6722175/372495b5-c7d4-4782-b33a-966658513de3)

### CODE REVIEW

File: React/CoreModules/RCTActionSheetManager.mm [hunk 1]
```diff
@@ -61,15 +69,10 @@ + (BOOL)requiresMainQueueSetup
...
   return NO;
 }
-#endif // [macOS]
 
 RCT_EXPORT_MODULE()
 
-#if !TARGET_OS_OSX // [macOS]
 @synthesize viewRegistry_DEPRECATED = _viewRegistry_DEPRECATED;
-#else // [macOS
-@synthesize bridge = _bridge;
```
Comment (harrieshin) [inline React/CoreModules/RCTActionSheetManager.mm:71]: didn't @amgleitman put this to resolve some bug before?

File: React/CoreModules/RCTActionSheetManager.mm [hunk 2]
```diff
@@ -37,22 +37,30 @@ @interface RCTActionSheetManager () <
...
+   * That selector no longer has has access to the method argument `callback`, so we must save it
+   * as an instance variable, that we can access in `menuItemDidTap`. We must do this as well for 
+   * `failureCallback` and `successCallback`.
+   */
   NSMapTable *_callbacks;
-  NSArray<NSSharingService*> *_excludedActivities;
-  NSString *_sharingSubject;
   RCTResponseSenderBlock _failureCallback;
   RCTResponseSenderBlock _successCallback;
+  NSArray<NSSharingService*> *_excludedActivities;
```
Comment (harrieshin) [inline React/CoreModules/RCTActionSheetManager.mm:48]: nit: why we moving lcoation?

File: React/CoreModules/RCTActionSheetManager.mm (hunk 1 above)
Comment (Saadnajmi) [reply React/CoreModules/RCTActionSheetManager.mm:71]: Normally you access the viewRegistry from the bridge. But the bridge is going away with the new arch, so my guess is this was an upstream refactor where `RCTActionSheetManager` has a direct reference to the view registry so you don't need a reference to the bridge. 

The view registry is used to look up what view to anchor the menu on. In other parts of this code, I updated the calls and things seem to work.

File: React/CoreModules/RCTActionSheetManager.mm (hunk 2 above)
Comment (Saadnajmi) [reply React/CoreModules/RCTActionSheetManager.mm:48]: The updated comment on lines 40-44 refer to all the callback ivars. So I wanted to group them together.

File: React/CoreModules/RCTActionSheetManager.mm [hunk 3]
```diff
@@ -198,14 +233,30 @@ - (void)presentViewController:(UIViewController *)alertController
       [actionButton setValue:cancelButtonTintColor forKey:@"titleTextColor"];
     }
     [alertController addAction:actionButton];
+#else // [macOS
+    if (index == cancelButtonIndex) {
+      // NSMenu doesn't need a cancel button, you can just click outside the menu
+      continue;
+    }
+
+    NSMenuItem *item = [[NSMenuItem alloc] initWithTitle:option action:@selector(menuItemDidTap:) keyEquivalent:@""];
```
Comment (harrieshin) [inline React/CoreModules/RCTActionSheetManager.mm:242]: if client was trying to share code between mac and ios and had stuff in the destructivebuttonindices, will it still show in this regular item? or am i reading this code wrong. who feeds the destructivebuttonindicies?

File: React/CoreModules/RCTActionSheetManager.mm (hunk 1 above)
Comment (harrieshin) [reply React/CoreModules/RCTActionSheetManager.mm:71]: i see you are trying to remove all instances like line 405?

File: React/CoreModules/RCTActionSheetManager.mm (hunk 3 above)
Comment (Saadnajmi) [reply React/CoreModules/RCTActionSheetManager.mm:242]: Menu items with destructive button indices will show up as regular menu items with no special styling. If it helps, I can update the test plan to show iOS and macOS side by side for the same tests.

File: React/CoreModules/RCTActionSheetManager.mm (hunk 1 above)
Comment (Saadnajmi) [reply React/CoreModules/RCTActionSheetManager.mm:71]: Yep, that matches what iOS is doing. I just copied the iOS logic there.

**Review by harrieshin — APPROVED** [review]

**Pull request merged**